unexpected ChannelClosedExceptions

(Feel free to suggest a better title. Just trying to capture the details on slack.)
(May or may not be related to #405.)

@jacob-koren reports:

```Client:
org.databene.contiperf.PerfTestExecutionError: java.lang.AssertionError: Error message from HttpResponseException not matches: 502--ChannelException at remote address: /10.0.3.170:22393 from service: 0.0.0.0/4140. Remote Info: Upstream Address: /10.0.5.108:21738, Upstream Client Id: Not Available, Downstream Address: /10.0.3.170:22393, Downstream Client Id: #/io.l5d.marathon/sre-testing, Trace Id: 6bc78e8f619f66f4.8633598551f506cf<:6bc78e8f619f66f4

    at org.databene.contiperf.util.ContiPerfUtil.executionError(ContiPerfUtil.java:58)
    at org.databene.contiperf.junit.JUnitInvoker.invoke(JUnitInvoker.java:46)
    at org.databene.contiperf.util.InvokerProxy.invoke(InvokerProxy.java:38)
    at org.databene.contiperf.PerformanceTracker.invoke(PerformanceTracker.java:87)
    at org.databene.contiperf.TimedRunner.run(TimedRunner.java:44)
    at java.lang.Thread.run(Thread.java:745)
Caused by: java.lang.AssertionError: Error message from HttpResponseException not matches: 502--ChannelException at remote address: /10.0.3.170:22393 from service: 0.0.0.0/4140. Remote Info: Upstream Address: /10.0.5.108:21738, Upstream Client Id: Not Available, Downstream Address: /10.0.3.170:22393, Downstream Client Id: #/io.l5d.marathon/sre-testing, Trace Id: 6bc78e8f619f66f4.8633598551f506cf<:6bc78e8f619f66f4
    at org.junit.Assert.fail(Assert.java:88)
    at org.junit.Assert.assertTrue(Assert.java:41)
    at com.zooz.smartrouting.utils.OperationsUtil.undeployAndGetToEnvAndAssert(OperationsUtil.java:77)
    at com.zooz.smartrouting.api.ConcurrentConfigurationOperationsFlow.runTest(ConcurrentConfigurationOperationsFlow.java:206)
    at com.zooz.smartrouting.api.ConcurrentConfigurationOperationsFlow.testDifferentFlowOwnerDeployingAndUndeployingInParallal(ConcurrentConfigurationOperationsFlow.java:172)
    at sun.reflect.GeneratedMethodAccessor18.invoke(Unknown Source)
    at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke(Method.java:483)
    at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
    at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
    at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
    at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
    at org.databene.contiperf.junit.JUnitInvoker.invoke(JUnitInvoker.java:43)
    ... 4 more

Server:
E 0615 15:35:20.500 THREAD25: service failure
com.twitter.finagle.ChannelClosedException: ChannelException at remote address: /10.0.3.170:22393 from service: 0.0.0.0/4140. Remote Info: Upstream Address: /10.0.5.108:21738, Upstream Client Id: Not Available, Downstream Address: /10.0.3.170:22393, Downstream Client Id: #/io.l5d.marathon/sre-testing, Trace Id: 6bc78e8f619f66f4.8633598551f506cf<:6bc78e8f619f66f4
    at com.twitter.finagle.NoStacktrace(Unknown Source)
```

I've been able to reproduce this locally:

- Run a [small web server that closes connections](https://gist.github.com/olix0r/872f206ed22fce058fcd67c7fdc52e65)
  - `go run shutter.go`
- run `./sbt linkerd-examples/http:run`
  - `echo "127.1 8080" >linkerd/examples/io.l5d.fs/ugly`
- in two terminals run: `while true ; do curl -H 'Host: ugly' localhost:4140 ; done`

After a few seconds, the following error is encountered in linkerd's logs:
```
om.twitter.finagle.ChannelClosedException: null at remote address: /127.0.0.1:8080 from service: 0.0.0.0/4140. Remote Info: Upstream Address: /127.0.0.1:62989, Upstream Client Id: Not Available, Downstream Address: /127.0.0.1:8080, Downstream Client Id: #/io.l5d.fs/ugly, Trace Id: 525ae9bf20f23725.e45f5414a27b8f9b<:525ae9bf20f23725
	at com.twitter.finagle.NoStacktrace(Unknown Source)
Caused by: java.nio.channels.ClosedChannelException
```

This should give us enough information to debug this in linkerd.  Thanks again for all your help documenting the problem, @jacob-koren.

--- a/CHANGES.md
+++ b/CHANGES.md
@@ -10,7 +10,7 @@
   * `l5d-dtab` now honored as a replacement for `dtab-local` as
     specified by users.
   * `l5d-dst-*` no longer set on responses
-
+* Fix graceful connection teardown on streaming HTTP responses #482.
 
 ## 0.6.0
 

--- a/linkerd/protocol/http/src/main/scala/com/twitter/finagle/buoyant/linkerd/DelayedRelease.scala
+++ b/linkerd/protocol/http/src/main/scala/com/twitter/finagle/buoyant/linkerd/DelayedRelease.scala
@@ -1,0 +1,16 @@
+package com.twitter.finagle
+package buoyant.linkerd
+
+import com.twitter.finagle.client.StackClient
+import com.twitter.finagle.http.{DelayedReleaseService, Request, Response}
+
+object DelayedRelease {
+
+  val module: Stackable[ServiceFactory[Request, Response]] =
+    new Stack.Module0[ServiceFactory[Request, Response]] {
+      val role = StackClient.Role.prepFactory
+      val description = "Prevents an HTTP service from being closed until its response completes"
+      def make(next: ServiceFactory[Request, Response]) =
+        next.map(new DelayedReleaseService(_))
+    }
+}

--- a/linkerd/protocol/http/src/main/scala/io/buoyant/linkerd/protocol/HttpConfig.scala
+++ b/linkerd/protocol/http/src/main/scala/io/buoyant/linkerd/protocol/HttpConfig.scala
@@ -16,24 +16,22 @@ class HttpInitializer extends ProtocolInitializer.Simple {
   protected type Req = com.twitter.finagle.http.Request
   protected type Rsp = com.twitter.finagle.http.Response
 
-  protected val defaultRouter = {
-    val pathStack = Http.router.pathStack
-      .prepend(Headers.Dst.PathFilter.module)
-      .replace(StackClient.Role.prepFactory, DelayedRelease.module)
-    val boundStack = Http.router.boundStack
-      .prepend(Headers.Dst.BoundFilter.module)
-    val clientStack = Http.router.clientStack
-      .prepend(http.AccessLogger.module)
-      .replace(HttpTraceInitializer.role, HttpTraceInitializer.clientModule)
-      .insertAfter(Retries.Role, http.StatusCodeStatsFilter.module)
-      .insertAfter(StackClient.Role.prepConn, Headers.Ctx.clientModule)
+  protected val pathStack = Http.router.pathStack
+    .prepend(Headers.Dst.PathFilter.module)
+    .replace(StackClient.Role.prepFactory, DelayedRelease.module)
+  protected val boundStack = Http.router.boundStack
+    .prepend(Headers.Dst.BoundFilter.module)
+  protected val clientStack = Http.router.clientStack
+    .prepend(http.AccessLogger.module)
+    .replace(HttpTraceInitializer.role, HttpTraceInitializer.clientModule)
+    .insertAfter(Retries.Role, http.StatusCodeStatsFilter.module)
+    .insertAfter(StackClient.Role.prepConn, Headers.Ctx.clientModule)
 
-    Http.router
-      .withPathStack(pathStack)
-      .withBoundStack(boundStack)
-      .withClientStack(clientStack)
-      .configured(RoutingFactory.DstPrefix(Path.Utf8(name)))
-  }
+  protected val defaultRouter = Http.router
+    .withPathStack(pathStack)
+    .withBoundStack(boundStack)
+    .withClientStack(clientStack)
+    .configured(RoutingFactory.DstPrefix(Path.Utf8(name)))
 
   protected val defaultServer = {
     val stk = Http.server.stack

--- a/linkerd/protocol/http/src/main/scala/io/buoyant/linkerd/protocol/HttpConfig.scala
+++ b/linkerd/protocol/http/src/main/scala/io/buoyant/linkerd/protocol/HttpConfig.scala
@@ -3,6 +3,7 @@ package protocol
 
 import com.fasterxml.jackson.annotation.JsonIgnore
 import com.twitter.finagle.{Path, Stack}
+import com.twitter.finagle.buoyant.linkerd.DelayedRelease
 import com.twitter.finagle.client.StackClient
 import com.twitter.finagle.buoyant.linkerd.{Headers, HttpTraceInitializer}
 import com.twitter.finagle.service.Retries
@@ -18,6 +19,7 @@ class HttpInitializer extends ProtocolInitializer.Simple {
   protected val defaultRouter = {
     val pathStack = Http.router.pathStack
       .prepend(Headers.Dst.PathFilter.module)
+      .replace(StackClient.Role.prepFactory, DelayedRelease.module)
     val boundStack = Http.router.boundStack
       .prepend(Headers.Dst.BoundFilter.module)
     val clientStack = Http.router.clientStack

--- a/linkerd/protocol/http/src/main/scala/io/buoyant/linkerd/protocol/HttpConfig.scala
+++ b/linkerd/protocol/http/src/main/scala/io/buoyant/linkerd/protocol/HttpConfig.scala
@@ -16,22 +16,24 @@ class HttpInitializer extends ProtocolInitializer.Simple {
   protected type Req = com.twitter.finagle.http.Request
   protected type Rsp = com.twitter.finagle.http.Response
 
-  protected val pathStack = Http.router.pathStack
-    .prepend(Headers.Dst.PathFilter.module)
-    .replace(StackClient.Role.prepFactory, DelayedRelease.module)
-  protected val boundStack = Http.router.boundStack
-    .prepend(Headers.Dst.BoundFilter.module)
-  protected val clientStack = Http.router.clientStack
-    .prepend(http.AccessLogger.module)
-    .replace(HttpTraceInitializer.role, HttpTraceInitializer.clientModule)
-    .insertAfter(Retries.Role, http.StatusCodeStatsFilter.module)
-    .insertAfter(StackClient.Role.prepConn, Headers.Ctx.clientModule)
+  protected val defaultRouter = {
+    val pathStack = Http.router.pathStack
+      .prepend(Headers.Dst.PathFilter.module)
+      .replace(StackClient.Role.prepFactory, DelayedRelease.module)
+    val boundStack = Http.router.boundStack
+      .prepend(Headers.Dst.BoundFilter.module)
+    val clientStack = Http.router.clientStack
+      .prepend(http.AccessLogger.module)
+      .replace(HttpTraceInitializer.role, HttpTraceInitializer.clientModule)
+      .insertAfter(Retries.Role, http.StatusCodeStatsFilter.module)
+      .insertAfter(StackClient.Role.prepConn, Headers.Ctx.clientModule)
 
-  protected val defaultRouter = Http.router
-    .withPathStack(pathStack)
-    .withBoundStack(boundStack)
-    .withClientStack(clientStack)
-    .configured(RoutingFactory.DstPrefix(Path.Utf8(name)))
+    Http.router
+      .withPathStack(pathStack)
+      .withBoundStack(boundStack)
+      .withClientStack(clientStack)
+      .configured(RoutingFactory.DstPrefix(Path.Utf8(name)))
+  }
 
   protected val defaultServer = {
     val stk = Http.server.stack

--- a/linkerd/protocol/http/src/test/scala/io/buoyant/linkerd/protocol/HttpInitializerTest.scala
+++ b/linkerd/protocol/http/src/test/scala/io/buoyant/linkerd/protocol/HttpInitializerTest.scala
@@ -1,0 +1,76 @@
+package io.buoyant.linkerd.protocol
+
+import com.twitter.finagle.{Service, ServiceFactory, Stack}
+import com.twitter.finagle.http.{Request, Response, Status, Version}
+import com.twitter.finagle.stack.nilStack
+import com.twitter.io.Reader
+import com.twitter.util.{Future, Promise, Time}
+import io.buoyant.test.Awaits
+import org.scalatest.FunSuite
+import org.scalatest.concurrent.Eventually
+import scala.language.reflectiveCalls
+
+class HttpInitializerTest extends FunSuite with Awaits with Eventually {
+
+  test("path stack: services are not closed until streams are complete") {
+    // Build a path stack that is controllable with promises (for only
+    // one request).
+    val serviceP, responseP, bodyP, respondingP, closedP = Promise[Unit]
+    val http = new HttpInitializer {
+      val svc = new Service[Request, Response] {
+        def apply(req: Request) = {
+          val rw = Reader.writable()
+          val rsp = Response(req.version, Status.Ok, rw)
+          val _ = bodyP.before(rw.close())
+          respondingP.setDone()
+          responseP.before(Future.value(rsp))
+        }
+
+        override def close(d: Time) = {
+          closedP.setDone()
+          Future.Unit
+        }
+      }
+      val sf = ServiceFactory { () => serviceP.before(Future.value(svc)) }
+
+      def make(params: Stack.Params = Stack.Params.empty) =
+        (pathStack ++ Stack.Leaf(Stack.Role("leaf"), sf)).make(params)
+    }
+
+    // The factory is returned immediately because it is wrapped in a
+    // FactoryToService.
+    val factory = http.make()
+    val svcf = factory()
+    assert(svcf.isDefined)
+    val svc = await(svcf)
+
+    // When a request is processed, first the service must be acquired
+    // from the service factroy, and then the response must be
+    // returned from the service.
+    val rspf = svc(Request())
+    assert(!rspf.isDefined)
+    assert(!respondingP.isDefined)
+
+    serviceP.setDone()
+    eventually { assert(respondingP.isDefined) }
+    assert(!rspf.isDefined)
+
+    responseP.setDone()
+    eventually { assert(rspf.isDefined) }
+
+    // Once the response is returned, FactoryToService tries to close
+    // the service factory. Ensure that the service is not closed
+    // until the response body is completely sent.
+    val rsp = await(rspf)
+    assert(rsp.isChunked)
+    assert(!closedP.isDefined)
+
+    // When the response body is written, it must be fully read from
+    // response before the service will be closed.
+    bodyP.setDone()
+    assert(!closedP.isDefined)
+
+    assert(await(rsp.reader.read(1)) == None)
+    eventually { assert(closedP.isDefined) }
+  }
+}

--- a/linkerd/protocol/http/src/test/scala/io/buoyant/linkerd/protocol/HttpInitializerTest.scala
+++ b/linkerd/protocol/http/src/test/scala/io/buoyant/linkerd/protocol/HttpInitializerTest.scala
@@ -1,10 +1,13 @@
 package io.buoyant.linkerd.protocol
 
-import com.twitter.finagle.{Service, ServiceFactory, Stack}
+import com.twitter.conversions.time._
+import com.twitter.finagle.{Service, ServiceFactory, Stack, param}
 import com.twitter.finagle.http.{Request, Response, Status, Version}
+import com.twitter.finagle.service.{Retries, RetryBudget}
 import com.twitter.finagle.stack.nilStack
 import com.twitter.io.Reader
-import com.twitter.util.{Future, Promise, Time}
+import com.twitter.util.{Future, MockTimer, Promise, Time}
+import io.buoyant.linkerd.protocol.http.ResponseClassifiers
 import io.buoyant.test.Awaits
 import org.scalatest.FunSuite
 import org.scalatest.concurrent.Eventually
@@ -13,8 +16,8 @@ import scala.language.reflectiveCalls
 class HttpInitializerTest extends FunSuite with Awaits with Eventually {
 
   test("path stack: services are not closed until streams are complete") {
-    // Build a path stack that is controllable with promises (for only
-    // one request).
+    // Build a path stack that is controllable with promises. This is
+    // only really useful for a single request.
     val serviceP, responseP, bodyP, respondingP, closedP = Promise[Unit]
     val http = new HttpInitializer {
       val svc = new Service[Request, Response] {
@@ -34,7 +37,7 @@ class HttpInitializerTest extends FunSuite with Awaits with Eventually {
       val sf = ServiceFactory { () => serviceP.before(Future.value(svc)) }
 
       def make(params: Stack.Params = Stack.Params.empty) =
-        (pathStack ++ Stack.Leaf(Stack.Role("leaf"), sf)).make(params)
+        (defaultRouter.pathStack ++ Stack.Leaf(Stack.Role("leaf"), sf)).make(params)
     }
 
     // The factory is returned immediately because it is wrapped in a
@@ -45,7 +48,7 @@ class HttpInitializerTest extends FunSuite with Awaits with Eventually {
     val svc = await(svcf)
 
     // When a request is processed, first the service must be acquired
-    // from the service factroy, and then the response must be
+    // from the service factory, and then the response must be
     // returned from the service.
     val rspf = svc(Request())
     assert(!rspf.isDefined)
@@ -72,5 +75,35 @@ class HttpInitializerTest extends FunSuite with Awaits with Eventually {
 
     assert(await(rsp.reader.read(1)) == None)
     eventually { assert(closedP.isDefined) }
+  }
+
+  test("path stack: retries") {
+    @volatile var requests = 0
+    val http = new HttpInitializer {
+      val sf = ServiceFactory.const(Service.mk[Request, Response] { req =>
+        requests += 1
+        Future.value(Response(req.version, Status.InternalServerError))
+      })
+
+      def make(params: Stack.Params = Stack.Params.empty) =
+        (defaultRouter.pathStack ++ Stack.Leaf(Stack.Role("leaf"), sf)).make(params)
+    }
+
+    val budget = RetryBudget(10.seconds, 0, 0.5)
+    val params = Stack.Params.empty +
+      param.ResponseClassifier(ResponseClassifiers.RetryableReadFailures) +
+      Retries.Budget(budget)
+    val factory = http.make(params)
+
+    val service = await(factory())
+
+    // First request just returns, since retry budget hasn't yet accrued.
+    val response0 = await(service(Request()))
+    assert(requests == 1)
+
+    // The second request is retryable because of the 50% retry
+    // budget.
+    val response1 = await(service(Request()))
+    assert(requests == 3)
   }
 }

--- a/router/core/src/main/scala/io/buoyant/router/Router.scala
+++ b/router/core/src/main/scala/io/buoyant/router/Router.scala
@@ -313,6 +313,7 @@ object StackRouter {
      */
     val stk = new StackBuilder[ServiceFactory[Req, Rsp]](stack.nilStack)
     stk.push(failureRecording)
+    stk.push(StackClient.Role.prepFactory, identity[ServiceFactory[Req, Rsp]](_))
     stk.push(factoryToService)
     stk.push(ClassifiedRetries.module)
     stk.push(StatsFilter.module)

